### PR TITLE
MSC2758: Proposal for a common identifier grammar

### PR DESCRIPTION
**Pull request by richvdh**
[Rendered](https://github.com/matrix-org/matrix-doc/blob/rav/proposals/textual_identifier_grammar/proposals/2758-textual-id-grammar.md)

### CODE REVIEW

File: proposals/2758-textual-id-grammar.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+ * [`capabilities`](https://matrix.org/docs/spec/client_server/r0.6.0#get-matrix-client-r0-capabilities)
+   identifiers.
+ * authentication types for the [User-Interactive Authentication mechanism](https://matrix.org/docs/spec/client_server/r0.6.0#user-interactive-authentication-api).
+ * login types for [`/_matrix/client/r0/login`](https://matrix.org/docs/spec/client_server/r0.6.0#post-matrix-client-r0-login).
+ * event types
+ * [`m.room.message` `msgtypes`](https://matrix.org/docs/spec/client_server/r0.6.0#m-room-message-msgtypes)
+ * `app_id` for [`POST /_matrix/client/r0/pushers/set`](https://matrix.org/docs/spec/client_server/r0.6.0#post-matrix-client-r0-pushers-set).
+ * `rule_ids`, `actions` and `tweaks` for [push rules](https://matrix.org/docs/spec/client_server/r0.6.0#push-rules).
+ * [E2E messaging algoritm names](https://matrix.org/docs/spec/client_server/r0.6.0#messaging-algorithm-names).
```
Comment (ara4n) [inline proposals/2758-textual-id-grammar.md:26]:
```suggestion
 * [E2E messaging algorithm names](https://matrix.org/docs/spec/client_server/r0.6.0#messaging-algorithm-names).
```

File: proposals/2758-textual-id-grammar.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+   example `com.example.identifier`.
+
+This grammar is intended for use entirely by internal identifiers, and *not*
+for user-visible strings.
+
+### Rationale
+
+ * Avoiding non-ascii characters sidesteps any issues with homoglyphs or
+   altenative encodings of the same characters.
+ * Avoiding upper-case character sidesteps any concerns over case-sensistivity.
```
Comment (ara4n) [inline proposals/2758-textual-id-grammar.md:54]:
```suggestion
 * Avoiding upper-case character sidesteps any concerns over case-sensitivity.
```

**Comment by ara4n** [comment]
lgtm

File: proposals/2758-textual-id-grammar.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+We define a "common namespaced identifier grammar". This can then be referenced
+by other parts of the grammar, in much the same way as [Unpadded
+Base64](https://matrix.org/docs/spec/appendices#unpadded-base64) is defined
+today.
+
+The grammar is defined as follows:
+
+ * The length of an identifier may not be less than one character or more than
+   255 characters.
```
Comment (KitsuneRal) [inline proposals/2758-textual-id-grammar.md:38]: Given that reserved identifiers start with `m.`, and unspecified identifiers should use reverse-DNS (aka Java Package Naming), what kind of identifiers can have length 1 or 2?

Comment (richvdh) [reply proposals/2758-textual-id-grammar.md:38]: > what kind of identifiers can have length 1 or 2?

good question. I think the best answer is "hopefully none, but in practice maybe some".  The difference as I see it is between "preferred" and "permitted": we would *prefer* that people use identifiers with length of at least 3, but APIs should *permit* them to use shorter ones.

File: proposals/2758-textual-id-grammar.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+ * Identifiers must start with one of the characters `[a-z]`, and be entirely
+   composed of the characters `[a-z]`, `[0-9]`, `-`, `_` and `.`.
+ * Identifiers starting with the characters `m.` are reserved for use by the
+   formal matrix specification.
+ * Implementations wishing to implement unspecified identifiers should follow
```
Comment (Sorunome) [inline proposals/2758-textual-id-grammar.md:43]: how about adding that prefixes must be separated by a `.` from the rest? Giving this example `com.example.identifierhithere` seems to be a valid identifier.

File: proposals/2758-textual-id-grammar.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+   the Java Package Naming conventions of using a reversed domain name, for
+   example `com.example.identifier`.
```
Comment (richvdh) [inline proposals/2758-textual-id-grammar.md:45]:
```suggestion

```
```suggestion
   the Java Package Naming convention of starting with a reversed domain
   name (with a dot after the domain name part). For example, for the 
   organisation `example.com`, a valid identifier would be
   `com.example.identifier`.
```

File: proposals/2758-textual-id-grammar.md (hunk 4 above)
Comment (richvdh) [reply proposals/2758-textual-id-grammar.md:43]: well, it *is* a valid identifier if you're `com.example` :).

Do you mean `com.exampleidentifier` is implied to be a valid identifier? Hopefully this is clarified now.

File: proposals/2758-textual-id-grammar.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+   255 characters.
+ * Identifiers must start with one of the characters `[a-z]`, and be entirely
+   composed of the characters `[a-z]`, `[0-9]`, `-`, `_` and `.`.
+ * Identifiers starting with the characters `m.` are reserved for use by the
+   formal matrix specification.
+ * Implementations wishing to implement unspecified identifiers should follow
+   the Java Package Naming convention of starting with a reversed domain
+   name (with a dot after the domain name part). For example, for the 
+   organisation `example.com`, a valid identifier would be
+   `com.example.identifier`.
```
Comment (turt2live) [inline proposals/2758-textual-id-grammar.md:47]: this is already covered by the proposal process? Specifically, https://matrix.org/docs/spec/proposals#early-release-of-a-msc-idea does not constrain implementations to having an MSC - it constrains implementations that have *ideas*, which is sufficiently vague to cover all unspecified behaviour.

Comment (richvdh) [reply proposals/2758-textual-id-grammar.md:47]: possibly. I feel like it's worth calling out in the spec as well, even if it says the same thing as the proposal process.

**Comment by richvdh** [comment]
Part of me feels like this is ready for FCP, but also I don't want to have a bunch of completed MSCs (or worse, stuff in the spec) which isn't actually used (yet). So I'm minded to hold off on it until we have a usecase for it which is also ready for FCP.

**Comment by turt2live** [comment]
We are in a position where we can run it against Widgets - we're in the spec PR stages for widgets, however its unique position allows for us to pile MSCs on top of the spec before it actually gets cut as a version.

**Comment by richvdh** [comment]
Now that we have things like #2790 building on this, I think we should FCP this

@mscbot fcp merge

File: proposals/2758-textual-id-grammar.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+We define a "common namespaced identifier grammar". This can then be referenced
+by other parts of the grammar, in much the same way as [Unpadded
+Base64](https://matrix.org/docs/spec/appendices#unpadded-base64) is defined
+today.
+
+The grammar is defined as follows:
+
+ * The length of an identifier may not be less than one character or more than
+   255 characters.
```
[the remark this answers is not in the document]
Comment (richvdh) [reply proposals/2758-textual-id-grammar.md]: fair. done.

**Comment by turt2live** [comment]
This merged 14 days ago 🎉

**Pull request merged**